18F/identity-idp
# Reimplement password strength as ViewComponent, custom element

## 🛠 Summary of changes

Refactors password strength UI element as a combination ViewComponent component + JavaScript custom element.

As part of this work, removes ad-hoc text size reduction previously applied to strength and feedback elements.

The refactoring adopts [frontend architecture conventions](https://github.com/18F/identity-idp/blob/main/docs/frontend.md), namely:

- TypeScript instead of (non-typechecked) JavaScript
- Custom elements instead of one-off packs
- ViewComponent component instead of Rails view partial
- BEM CSS class naming without abbreviations instead of nonstandard class names (see [related comment](https://github.com/18F/identity-idp/pull/9811#discussion_r1432990205))
- Component-specific stylesheets instead of main stylesheet include
- Dependencies defined within workspace packages

Related: #9811

**Why?**

- Simplify usage and ensures consistency by embedding password strength as a subcomponent of the existing password confirmation experience, to 
- Improves maintainability by making it easier to test the custom element, which has more controllable side-effects than a JavaScript pack
- Reduces size of main application stylesheet to improve load times in critical paths
- Aligns text styling to current Figma designs
- Improves accessibility and usability by increasing the font size of text feedback

## 📜 Testing Plan

Verify there are no regressions in the behavior of password strength:

1. Go to http://localhost:3000
2. Sign in
3. On account dashboard, click Edit Password
4. (Reauthenticate if prompted)
5. Enter a new password
6. Observe password feedback

## 👀 Screenshots

Before|After
---|---
![image](https://github.com/18F/identity-idp/assets/1779930/3ee20ec0-77e0-4900-89a6-9cc802a9e41d)|![image](https://github.com/18F/identity-idp/assets/1779930/98f4a315-c72a-46cf-8696-711b58f2c488)
